logprep/processor/calculator/processor.py

```diff
@@ -0,0 +1,76 @@
+"""
+Calculator
+==========
+
+The Calculator can be used to calculate with or without field values.
+For further information for the rule language see: :ref:`calulator_rule`
+
+"""
+from functools import partial
+import re
+
+from pyparsing import ParseException
+
+from logprep.abc import Processor
+from logprep.processor.base.exceptions import DuplicationError
+from logprep.util.helper import add_field_to, get_dotted_field_value
+
+from .fourFn import BNF, evaluate_stack, exprStack
+from .rule import CalculatorRule
+
+
+class Calculator(Processor):
+    """A Processor to calculate with and without field values"""
+
+    rule_class = CalculatorRule
+
+    def _apply_rules(self, event, rule):
+        source_fields = rule.source_fields
+        source_field_values = map(partial(get_dotted_field_value, event), source_fields)
+        source_field_dict = dict(zip(source_fields, source_field_values))
+        self._check_for_missing_values(event, rule, source_field_dict)
+        expression = self._template(rule.calc, source_field_dict)
+        result = self._calculate(event, rule, expression)
+        self._write_target_field(event, rule, result)
+
+    @staticmethod
+    def _template(string: str, source: dict) -> str:
+        for key, value in source.items():
+            key = key.replace(".", r"\.")
+            pattern = r"\$\{(" + rf"{key}" + r")\}"
+            string = re.sub(pattern, str(value), string)
+        return string
+
+    def _calculate(self, event, rule, expression):
+        try:
+            _ = BNF().parseString(expression, parseAll=True)
+            result = evaluate_stack(exprStack[:])
```

Comment:
I am not really a fan of this magically imported `exprStack`. It's not easily understandable where it comes from and why it suddenly holds the expression that needs evaluation. Maybe we can rewrite this fourFn example and turn it into a proper class that is closed in it self.

Reply:
Done